### PR TITLE
Delegated authentication - with newer screens

### PR DESCRIPTION
**Pull request by robertdalmeida**
<changed> Support for delegated authentication  </changed>

## Summary

Delegated authentication screens

- [x] create authentication sdk release.
- [x] Ready to merge

### Register 

https://github.com/Adyen/adyen-ios/assets/96450350/4ffef87c-5eaf-43a1-a952-356cfe4d2797

### Approval

https://github.com/Adyen/adyen-ios/assets/96450350/c5b1ba24-c85f-49c6-a0a2-8b2d656b046e

### Delete

https://github.com/Adyen/adyen-ios/assets/96450350/ce691546-7e0e-4be5-a24f-58f3680381e0

### Registration Error

https://github.com/Adyen/adyen-ios/assets/96450350/d7076010-39ac-4a7b-9763-8cd15ccddfd5

### Error on approve

https://github.com/Adyen/adyen-ios/assets/96450350/1ab2be1b-d1c0-47b7-a351-4e4b2ad7e32f

### CODE REVIEW

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 1]
```diff
@@ -88,69 +83,228 @@
...
+        }
+        
+        private func showApprovalScreen(
+            delegatedAuthenticationInput: String,
+            cardType: CardType?,
+            cardNumber: String?,
+            completion: @escaping (Result<(daOutput: String, delete: Bool?), ApprovalFlowError>
+            ) -> Void
+        ) {
+            presenter.showApprovalScreen(
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift:193]: There is quite a lot of indentation going on. Do you think we could split it up in multiple methods?

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift [hunk 2]
```diff
@@ -88,69 +83,228 @@
...
+                                     guard let self else { return }
+                                     self.presenter.showAuthenticationError(component: self) {
+                                         completion(.failure(.authenticationServiceFailed(underlyingError: error)))
+                                     }
+                                 })
+                },
+                fallbackHandler: {
+                    completion(.failure(.fallbackTo3ds))
+                },
+                removeCredentialsHandler: { [weak self] in
```
Comment (goergisn) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift:205]: Same here (Could be that it's just Github formatting but maybe splitting would also improve readability)

**Review by goergisn — DISMISSED** [review]

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift (hunk 1 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift:193]: Sure. You know these would look much better with async 🤞. 
For now i've moved some of them in a separate method. 
also you would notice a bit more indentation than usual - due to below code right on top.
```
#if canImport(AdyenAuthentication)
    @_spi(AdyenInternal) import Adyen
```

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift (hunk 2 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift:205]: https://github.com/Adyen/adyen-ios/pull/1726#discussion_r1665801625

**Review by goergisn — DISMISSED** [review]

File: AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDAScreenPresenter.swift [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    func showRegistrationError(component: Component, handler: @escaping () -> Void)
+    func showDeletionConfirmation(component: Component, handler: @escaping () -> Void)
+
+    var presentationDelegate: PresentationDelegate? { get set }
+}
+
+/// This type handles the presenting of the Delegate authentication screens of Register and Approval.
+@available(iOS 16.0, *)
+internal final class ThreeDS2PlusDAScreenPresenter: ThreeDS2PlusDAScreenPresenterProtocol {
+    /// Delegates `PresentableComponent`'s presentation.
```
Comment (erenbesel) [inline AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDAScreenPresenter.swift]: This comment is misplaced it seems. For `presentationDelegate` maybe?

File: AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift [hunk 4]
```diff
@@ -48,10 +50,13 @@ internal func createDefaultThreeDS2CoreActionHandler(
     delegatedAuthenticationConfiguration: ThreeDS2Component.Configuration.DelegatedAuthentication?
 ) -> AnyThreeDS2CoreActionHandler {
     #if canImport(AdyenAuthentication)
-        if #available(iOS 14.0, *), let delegatedAuthenticationConfiguration {
-            return ThreeDS2PlusDACoreActionHandler(context: context,
-                                                   appearanceConfiguration: appearanceConfiguration,
-                                                   delegatedAuthenticationConfiguration: delegatedAuthenticationConfiguration)
+        if #available(iOS 16.0, *), let delegatedAuthenticationConfiguration {
+            return (try? ThreeDS2PlusDACoreActionHandler(context: context,
```
Comment (erenbesel) [inline AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift]: nitpicky a bit but these lines are too long to nil-coalesce I think

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        approvalView.descriptionLabel.text = localizedString(.threeds2DAApprovalDescription, localizationParameters)
+        approvalView.firstButton.title = localizedString(.threeds2DAApprovalPositiveButton, localizationParameters)
+        approvalView.secondButton.title = localizedString(.threeds2DAApprovalNegativeButton, localizationParameters)
+        approvalView.additionalInformationStackView.isHidden = true
+
+        switch (amount, cardNumber, cardType) {
+        case (.none, .none, .none), (.none, _, .none), (.none, .none, _):
+            approvalView.cardAndAmountDetailsStackView.isHidden = true
+        
+        case let (amount, cardNumber, cardType):
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift]: If I'm understanding correctly this switch shouldn't be needed. The only case for unhiding the view is if `cardNumber` and `cardType` are non-nil. Could we just check that with an `If` and achieve the same?

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                localizedString(.threeds2DARegistrationErrorTitle, localizationParameters)
+            case let .deletionConfirmation(localizationParameters):
+                localizedString(.threeds2DADeletionConfirmationTitle, localizationParameters)
+            }
+        }
+        
+        internal var image: UIImage? {
+            switch self {
+            case .authenticationFailed:
+                UIImage(named: "feedback", in: Bundle.actionsInternalResources, compatibleWith: nil)
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift]: Could we put the strings in a private constant enum here

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                                   renderingMode: .alwaysOriginal)
+            case .deletionConfirmation:
+                UIImage(systemName: "trash",
+                        withConfiguration: UIImage.SymbolConfiguration(weight: .ultraLight))?
+                    .withTintColor(.systemGray,
+                                   renderingMode: .alwaysOriginal)
+            }
+        }
+        
+        internal var message: String {
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift:53]: A bit hard to understand from the tests but do we have tests covering all these properties to make sure we have the correct values? If not, let's have tests covering `all` these conditional return values.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationErrorView.swift [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    // MARK: - Configuration
+    
+    private func configureViews() {
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        scrollView.addSubview(contentStackView)
+        addSubview(scrollView)
+        addSubview(buttonsStackView)
+        
+        NSLayoutConstraint.activate([
+            scrollView.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor, constant: 30),
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationErrorView.swift]: Many magic numbers, let's put into a constants enum

**Review by erenbesel — COMMENTED** [review]
Great one Robert. Are the changes to our action files (test-spm, cocoapods etc) coming from you or another branch you copied from? Couldn't fully get why those files need to be changed

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift (hunk 5 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift]: At the time writing - it seemed readable. But i get your point. Made it to an if.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift (hunk 7 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/DAErrorViewController.swift:53]: I've updated the tests to cover these cases, also renamed the test names to hopefully be more readable.

**Comment by robertdalmeida** [comment]
> Great one Robert. Are the changes to our action files (test-spm, cocoapods etc) coming from you or another branch you copied from? Couldn't fully get why those files need to be changed

The changes to the github actions are intentional. 
1. They contain the changes needed for consuming the authentcation sdk version 3.0.0
2. Using a newer version of mac os as the older version of xcode used to run the test for the cocoaPod action doesn't support all the values of `LABiometryType` 
    a. The newer version of swift doesn't support 'generate-xcodeproj` => XCodeGen

File: AdyenActions/Components/3DS2/ThreeDS2Component.swift [hunk 9]
```diff
@@ -49,24 +54,28 @@ public final class ThreeDS2Component: ActionComponent {
...
-            /// - Parameter localizedAuthenticationReason: The localized reason string show to the user while authentication flow.
-            /// - Parameter appleTeamIdentifier: The Apple registered development team identifier.
-            public init(localizedRegistrationReason: String, localizedAuthenticationReason: String, appleTeamIdentifier: String) {
-                self.localizedRegistrationReason = localizedRegistrationReason
-                self.localizedAuthenticationReason = localizedAuthenticationReason
-                self.appleTeamIdentifier = appleTeamIdentifier
+            /// - Parameter relyingPartyIdentifier: The relying party identifier that is used for PassKeys
+            /// - Parameter delegatedAuthenticationComponentStyle: The delegated authentication component style.
+            /// - Parameter localizationParameters: The localization parameters, leave it nil to use the default parameters.
+            public init(relyingPartyIdentifier: String,
```
Comment (erenbesel) [inline AdyenActions/Components/3DS2/ThreeDS2Component.swift:73]: This struct is deep in the hierarchy but nonetheless the init is public and it's changed. If (albeit very unlikely) anyone is using the init directly, it's gonna be a breaking change for them.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+            subView.layer.masksToBounds = true
+            subView.clipsToBounds = true
+        }
+        return stackView
+    }
+}
+
+@available(iOS 16.0, *)
+extension UILabel {
+    internal static func make(style: TextStyle, accessibilityPostfix: String, multiline: Bool = false) -> UILabel {
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift]: For all these extensions here:
- Why not init but a static function?
- Why is there a need for iOS 16?
- Let's move them to a separate extension file.

File: Demo/Configuration.swift [hunk 11]
```diff
@@ -65,10 +65,7 @@ internal enum ConfigurationConstants {
     ]]
     
     static var delegatedAuthenticationConfigurations: ThreeDS2Component.Configuration.DelegatedAuthentication {
-        .init(localizedRegistrationReason: "Authenticate your card!",
-              localizedAuthenticationReason: "Register this device!",
-              appleTeamIdentifier: appleTeamIdentifier)
-        
+        .init(relyingPartyIdentifier: "test-authentication-adyen.netlify.app")
```
Comment (erenbesel) [inline Demo/Configuration.swift:68]: Case in point: Creation of the config struct changed

File: AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift [hunk 12]
```diff
@@ -48,7 +50,7 @@ internal func createDefaultThreeDS2CoreActionHandler(
     delegatedAuthenticationConfiguration: ThreeDS2Component.Configuration.DelegatedAuthentication?
 ) -> AnyThreeDS2CoreActionHandler {
     #if canImport(AdyenAuthentication)
-        if #available(iOS 14.0, *), let delegatedAuthenticationConfiguration {
+        if #available(iOS 16.0, *), let delegatedAuthenticationConfiguration {
             return ThreeDS2PlusDACoreActionHandler(context: context,
```
Comment (erenbesel) [inline AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift:54]: So previous `??` check was not needed then? Seemed like there was some checking and returning a different one etc

**Review by erenbesel — COMMENTED** [review]
Some more comments :)

File: AdyenActions/Components/3DS2/ThreeDS2Component.swift (hunk 9 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/ThreeDS2Component.swift:73]: Yes in the strictest sense the public api has to change. 
Though we know that Delegated authentication is not used by anyone as yet, which will change in sometime.

File: Demo/Configuration.swift (hunk 11 above)
Comment (robertdalmeida) [reply Demo/Configuration.swift:68]: https://github.com/Adyen/adyen-ios/pull/1726#discussion_r1673530272

File: AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift (hunk 12 above)
Comment (robertdalmeida) [reply AdyenActions/Components/3DS2/Action handlers/AnyThreeDS2ActionHandler.swift:54]: It wasn't needed indeed. It was assuming that initializer for the AuthenticationService(Authentication framework) was throwing, but in the latest release it is not throwing. So there is no fall back required at this point.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift (hunk 10 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/DelegatedAuthenticationView.swift]: I've moved them in separate files. Please have a look.

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/ViewHelpers/UIStackViewHelper.swift [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        self.spacing = spacing
+        self.distribution = distribution
+        self.translatesAutoresizingMaskIntoConstraints = false
+        
+        if withBackground {
+            let subView = UIView(frame: view.bounds)
+            if #available(iOS 13.0, *) {
+                subView.backgroundColor = UIColor.secondarySystemBackground
+            } else {
+                // Fallback on earlier versions
```
Comment (goergisn) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/ViewHelpers/UIStackViewHelper.swift]: The else branch could be omitted

File: Scripts/test-SPM-integration.sh [hunk 14]
```diff
@@ -61,20 +62,22 @@ let package = Package(
...
 " > Package.swift
 
+
+echo '############# swift package update ###############'
 swift package update
 
 # Build for generic iOS device
 echo '############# Build for generic iOS device ###############'
-xcodebuild build -scheme TempProject -destination 'generic/platform=iOS' -skipPackagePluginValidation -quiet
+xcodebuild build -scheme TempProject -destination 'generic/platform=iOS' -skipPackagePluginValidation -quiet -derivedDataPath ./.dd
```
Comment (goergisn) [inline Scripts/test-SPM-integration.sh:71]: Why specifying the derived data path? It defaults to .build right?

File: Scripts/test-CocoaPods-integration.sh [hunk 15]
```diff
@@ -41,41 +41,43 @@ rm -rf $PROJECT_NAME
...
-    ]
-)
-
-" > Package.swift
-
-swift package update
-
-swift package generate-xcodeproj
+echo "
+name: $PROJECT_NAME
```
Comment (goergisn) [inline Scripts/test-CocoaPods-integration.sh:45]: I like this change - did you already have this code somewhere?

File: .github/workflows/test_cocoapods_integration.yml [hunk 16]
```diff
@@ -11,7 +11,7 @@ on:
 jobs:
 
   pods:
-    runs-on: macos-12-xl
+    runs-on: macos-14-xlarge
```
Comment (goergisn) [inline .github/workflows/test_cocoapods_integration.yml:14]: Is this update still needed? (I agree that it makes sense to upgrade but was wondering if it's required - did you run the compatibility tests already?)

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/ViewHelpers/UIStackViewHelper.swift (hunk 13 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/ViewHelpers/UIStackViewHelper.swift]: Sure. Thanks.

File: Scripts/test-SPM-integration.sh (hunk 14 above)
Comment (robertdalmeida) [reply Scripts/test-SPM-integration.sh:71]: The path defaults to what is set by default in the system. 
`.build` is for spm. 

I had an issue with multiple runs as xcode wasn't able to delete the previous run state information on a failed run and i was getting some inconsistent results only on the CI.
Maintaining the derived data in this temp folder fixed it as there is a nicely set `trap` to clean up on this script. so each run was on a clean slate. 

Let me know if you don't want it here. Happy to revert it.

File: Scripts/test-CocoaPods-integration.sh (hunk 15 above)
Comment (robertdalmeida) [reply Scripts/test-CocoaPods-integration.sh:45]: I've used the same script from the existing `test-SPM.sh`

File: .github/workflows/test_cocoapods_integration.yml (hunk 16 above)
Comment (robertdalmeida) [reply .github/workflows/test_cocoapods_integration.yml:14]: Strictly - I think it is not needed anymore, as we have downgraded the code. 
But if you compare with the other actions, it has the same configuration we are using in `test-SPM-integration.yml`

Compatibility tests - passed yesterday. 

let me know if you need to change/revert this. Happy to.

**Review by goergisn — DISMISSED** [review]

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+                                         style: .cancel)
+        alertController.addAction(cancelAction)
+        alertController.addAction(fallbackAction)
+        alertController.addAction(removeAction)
+        return alertController
+    }()
+
+    private lazy var approvalView: DelegatedAuthenticationView = .init(style: style)
+    
+    internal typealias Handler = () -> Void
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift]: I see this defined 3 times, maybe move it into the DA module or actions module (as internal access still)? We can name it `VoidHandler`

**Review by erenbesel — DISMISSED** [review]
Thanks a lot Robert! Approved with a tiny suggestion

File: AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift (hunk 17 above)
Comment (robertdalmeida) [reply AdyenActions/UI/View Controllers/DelegatedAuthentication/DAApprovalViewController.swift]: Sure. Done. Thanks.

**Review by erenbesel — APPROVED** [review]
🎉

**Review by goergisn — APPROVED** [review]

**Pull request merged**